packages/core/crates/core-misc/src/environment.rs

```diff
@@ -0,0 +1,265 @@
+
+    network.id = environment.network_id.to_owned();
+
+    if let Some(custom_provider) = maybe_custom_provider {
+        network.default_provider = String::from(custom_provider);
+    }
+
+    match satisfies(version, environment.version_range.to_owned()) {
+        Ok(true) => Ok(ResolvedEnvironment {
+            id: String::from(environment_id),
+            network: network.to_owned(),
+            environment_type: environment.environment_type,
+            channel_contract_deploy_block: environment.channel_contract_deploy_block,
+            token_contract_address: environment.token_contract_address.to_owned(),
+            channels_contract_address: environment.channels_contract_address.to_owned(),
+            xhopr_contract_address: environment.xhopr_contract_address.to_owned(),
+            boost_contract_address: environment.boost_contract_address.to_owned(),
+            stake_contract_address: environment.stake_contract_address.to_owned(),
+            network_registry_contract_address: environment
+                .network_registry_contract_address
+                .to_owned(),
+            network_registry_proxy_contract_address: environment
+                .network_registry_proxy_contract_address
+                .to_owned(),
+        }),
+        Ok(false) => match supported_environments(mono_repo_path) {
+            Ok(envs) => Err(JsValue::from(format!(
+                "environment {} is not supported, supported environments {:?}",
+                environment_id,
+                envs.iter()
+                    .map(|e| e.id.to_owned())
+                    .collect::<Vec<String>>()
+                    .join(", ")
+            ))),
+            Err(e) => Err(e),
+        },
+        Err(e) => Err(e),
+    }
+}
+
+pub mod wasm {
+    use wasm_bindgen::prelude::*;
+    use wasm_bindgen::JsValue;
+
+    macro_rules! clean_mono_repo_path {
+        ($v:expr,$r:ident) => {
+            let $r = $v.strip_suffix("/").unwrap_or($v);
+        };
+    }
+
+    #[wasm_bindgen]
+    pub fn supported_environments(mono_repo_path: &str) -> Result<JsValue, JsValue> {
+        clean_mono_repo_path!(mono_repo_path, cleaned_mono_repo_path);
+
+        let supported_envs = super::supported_environments(cleaned_mono_repo_path)?;
+
+        serde_wasm_bindgen::to_value(&supported_envs).map_err(|e| JsValue::from(e.to_string()))
+    }
+
+    #[wasm_bindgen]
+    pub fn resolve_environment(
+        mono_repo_path: &str,
+        environment_id: &str,
+        maybe_custom_provider: Option<String>,
+    ) -> Result<JsValue, JsValue> {
+        clean_mono_repo_path!(mono_repo_path, cleaned_mono_repo_path);
+
+        let resolved_environment = super::resolve_environment(
+            cleaned_mono_repo_path,
+            environment_id,
+            maybe_custom_provider.as_ref().map(|c| c.as_str()),
+        )?;
+
+        serde_wasm_bindgen::to_value(&resolved_environment)
+            .map_err(|e| JsValue::from(e.to_string()))
```

Comment:
Because `|e| JsValue::from(e.to_string())` repeats a lot, you can define `pub fn as_jsvalue<T>(v: T) -> JsValue where T: Display { v.to_string().into() }` and just do `x.map_err(as_jsvalue)`

Reply:
why don't use a macro ;-) ?